common/kernel/timing.cc

```diff
@@ -915,23 +1065,29 @@ CriticalPath TimingAnalyser::build_critical_path_report(domain_id_t domain_pair,
         seg_route.net = net->name;
         report.segments.push_back(seg_route);
 
-        last_cell = sink_cell;
-        last_port = sink.port;
+        prev_cell = sink_cell;
+        prev_port = sink.port;
+        is_startpoint = false;
     }
 
-    int clockCount = 0;
-    auto sinkClass = ctx->getPortTimingClass(crit_path.back().cell, crit_path.back().port, clockCount);
-    if (sinkClass == TMG_REGISTER_INPUT && clockCount > 0) {
-        auto sinkClockInfo = ctx->getPortClockingInfo(crit_path.back().cell, crit_path.back().port, 0);
-        auto setup = sinkClockInfo.setup;
-
+    // Add setup/hold time as final segment
+    // And add hold time as min bound
+    if (register_end) {
         CriticalPath::Segment seg_logic;
-        seg_logic.type = CriticalPath::Segment::Type::SETUP;
-        seg_logic.delay = setup;
-        seg_logic.from = std::make_pair(last_cell->name, last_port);
+        seg_logic.delay = DelayPair(0);
+        if (longest_path) {
+            seg_logic.type = CriticalPath::Segment::Type::SETUP;
+            seg_logic.delay += ep_clk_info.setup;
+        } else {
+            seg_logic.type = CriticalPath::Segment::Type::HOLD;
+            seg_logic.delay -= ep_clk_info.hold;
+        }
+        seg_logic.from = std::make_pair(prev_cell->name, prev_port);
         seg_logic.to = seg_logic.from;
         seg_logic.net = IdString();
         report.segments.push_back(seg_logic);
+
+        report.bound.min_delay = ep_clk_info.hold.min_delay;
```

Comment:
Oops this is wrong now. A hold time violation has negative delay now with the negative segments.